package com.veyzo.media_editing_system.Service;

import com.veyzo.media_editing_system.Model.User;
import com.veyzo.media_editing_system.Repository.UserRepository;
import com.veyzo.media_editing_system.dto.request.LoginRequest;
import com.veyzo.media_editing_system.dto.request.RegisterRequest;
import com.veyzo.media_editing_system.dto.response.AuthResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JWTService jwtService;


    public User register(RegisterRequest req){
        if(userRepository.existsByEmail(req.email()))
            throw new RuntimeException("Bu e-posta zaten kayıtlı.");
        if(userRepository.existsByUsername(req.username()))
            throw new RuntimeException("Bu kullanıcı adı zaten kayıtlı.");

        return userRepository.save(
                User.builder()
                        .username(req.username())
                        .email(req.email())
                        .password(passwordEncoder.encode(req.password()))
                        .build()
        );
    }

    public AuthResponse login(LoginRequest req){
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(req.email(), req.password())
        );
        User user = userRepository.findByEmail(req.email())
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı."));

        String jwtToken = jwtService.generateToken(user);
        return new AuthResponse(jwtToken);
    }
}
